language/move-analyzer/src/symbols.rs

```diff
@@ -0,0 +1,2184 @@
+        for (pos, module_ident, module_def) in modules {
+            let (defs, symbols) = Self::get_mod_outer_defs(
+                &pos,
+                &mut references,
+                module_def,
+                &files,
+                &file_id_mapping,
+            );
+            mod_outer_defs.insert(*module_ident, defs);
+            mod_use_defs.insert(*module_ident, symbols);
+            match source_files.get(&pos.file_hash()) {
+                Some((fpath, _)) => {
+                    mod_ident_map.insert(
+                        // if canonicalization fails, simply use "regular" path to continue
+                        // processing
+                        fs::canonicalize(fpath.as_str())
+                            .unwrap_or_else(|_| PathBuf::from(fpath.as_str())),
+                        *module_ident,
+                    );
+                }
+                None => continue,
+            }
+        }
+
+        let mut symbolicator = Symbolicator {
+            mod_outer_defs,
+            files,
+            file_id_mapping,
+            type_params: Scope::new(),
+            current_mod: Option::None,
+        };
+
+        for (_, module_ident, module_def) in modules {
+            let use_defs = mod_use_defs.get_mut(module_ident).unwrap();
+            symbolicator.current_mod = Some(*module_ident);
+            symbolicator.mod_symbols(module_def, &mut references, use_defs);
+        }
+
+        Ok(Symbols {
+            references,
+            mod_use_defs,
+            mod_ident_map,
+            file_name_mapping,
+        })
+    }
+
+    /// Get empty symbols
+    pub fn empty_symbols() -> Symbols {
+        Symbols {
+            mod_use_defs: BTreeMap::new(),
+            mod_ident_map: BTreeMap::new(),
+            references: BTreeMap::new(),
+            file_name_mapping: BTreeMap::new(),
+        }
+    }
+
+    /// Main AST traversal functions
+
+    /// Get symbols for outer definitions in the module (functions, structs, and consts)
+    fn get_mod_outer_defs(
+        loc: &Loc,
+        references: &mut BTreeMap<DefLoc, BTreeSet<UseLoc>>,
+        mod_def: &ModuleDefinition,
+        files: &SimpleFiles<Symbol, String>,
+        file_id_mapping: &BTreeMap<FileHash, usize>,
+    ) -> (ModuleDefs, UseDefMap) {
+        let mut structs = BTreeMap::new();
+        let mut constants = BTreeMap::new();
+        let mut functions = BTreeMap::new();
+        let mut use_def_map = UseDefMap::new();
+
+        for (pos, name, def) in &mod_def.structs {
+            match Self::get_start_loc(&pos, files, file_id_mapping) {
+                Some(name_start) => {
+                    let mut field_defs = vec![];
+                    if let StructFields::Defined(fields) = &def.fields {
+                        for (fpos, fname, _) in fields {
+                            match Self::get_start_loc(&fpos, files, file_id_mapping) {
+                                Some(start) => {
```

Comment:
Same sort of things as above, trying to do a match where you do `let data = match ... { ... None => continue }` is going to make this a lot more readable